maven-artifact/src/test/java/org/apache/maven/artifact/ArtifactTest.java

```diff
@@ -0,0 +1,43 @@
+package org.apache.maven.artifact;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.File;
+import java.nio.file.Path;
+
+import org.junit.jupiter.api.Test;
+
+public class ArtifactTest {
+
+    @Test
+    public void testPathYieldsFile() {
+        final Artifact artifact = spy( Artifact.class );
```

Comment:
This is testing implementation. Spies aren't needed here. Just test that calling setPath changes the return value of getFile and vice versa. And you can use actual Paths/Files instead of mocks.

Reply:
This seems to me like personal preference. Actually the Spy is used to test the default implementation, which is a pretty valid use of spies.

Reply:
This code *is* creating actual Paths/Files and no mocks of those classes *explicitly*. It is the implementation's decision how it works internally, so it is even fine that Mockito creates mocks internally.

Reply:
Agree using mocks is like not writing a test so if possible just use impls and temporaryfolder rule, this is enough IMHO

Reply:
Why is this like "not writing a test"? Can you please elaborate on that? In fact, the test is perfectly well and proofs exactly what it should, without the need of implementing lots and lots of unused methods.

Reply:
Right but the mock is useless and you test the mock not prod code, this is what i meant by useless. It costs nothing to write a test using 100% of prod code there, so it is worth it imho if we add a test for that addition.

Reply:
This is not true, I am not testing the mock at all, and it is anything but useless: I am testing the *original* default methods of the interface, as this is a *spy* (if it would be a *mock*, then the test would *fail* -- maybe you did not read the actual code lines I wrote?), I use the mocked methods just for fulfilling the interface without the need of creating an *unnecessary* explicit implementation class. And "nothing" is far from the truth: The *spy* (it is *not* a mock) costed me exactly one minute, while implementing an real Artifact costs me much more time, provides *useless* code lines, and forces future maintenance of them!

In fact, I wrote that code *by intention* as it is *exactly* what I want to proof: Is the *default* method of the *interface* working correctly? I do *not* want to test if an *actual implementation* is working correctly!

Reply:
Apparently some of you misunderstood what this code does. So let me explain with comments:
```java
final Artifact artifact = spy( Artifact.class ); // if you have ANY implementation that does NOT overwrite the default methods
artifact.setPath( Paths.get( "testPath" ) ); // when invoking setPATH
verify( artifact ).setFile( new File( "testPath" ) ); // then getFILE must get invoked in turn with exactly the same argument
when( artifact.getFile() ).thenReturn( new File( "testFile" ) );
assertEquals( Paths.get( "testFile" ), artifact.getPath() ); // when someone call getPATH it must get taken from getFILE (see line above)
```

Reply:
I get that but you precisely do what i say: not test what you think and test the mock.
To prove it, do a proxy based test, it will fail.
Personally i think the defauklt artifact test is more than sufficient since this code must never be used in practise but it is important to add useful tests validating prod code IMHO.

Reply:
So actually it is OK if there is *no* test for the *default methods*?

Reply:
Even for @elharo ?

Reply:
I might not be seeing it, but I don't think this is doing what you claim it is doing. A test for the default method would:

1. create a class that does not override the default method
2. call setPath on that class
3. verify that getPath returns the new value

It does not need mocks and it certainly does not need spies.

Reply:
@elharo So in contrast to @rmannibucau you do want to keep the test for the default methods, or why else do you continue this thread?

So for clarity:
1. The result of calling `spy()` *is exactly* a class that implements the interface but does not override the default methods.
2. The test *does* call `setPath` on that class
3. No, the test *shall not* test whether `getPath` returns what `setPath` has set, as it is *not* a test for some arbitrary implementation, but is a test for *the default method* itself and ONLY for itself. This IS how pure interface get tested, as otherwise you would NOT test the interface but test THE SUBCLASS's ability to store between get/set, which is NOT the intention of the test. It actually shall proof that `setPath` effectively calls `setFile`, as it tests the default methods ONLY, and that is exactly what the default methods shall do, and that is exactly what the test does, and that is why the test is named `testPathYieldsFile` but NOT `testGetPathYieldsSetPath`.

Reply:
Side note: 1 is not strictly true, it overrides it even if it delegates it at the end with mock handlers so not exactly the same.

Reply:
...yes, but which is completely irrelevant for the sake of this actual test!

Reply:
it does not test the default method so I think it is relevant if @elharo insists to keep the test, the use case he described is the only one to test about the interface and is not tested by the test you wrote.

Reply:
This is not true. It *does* implement the default method. There is a simple proof: Change the default implementation of `setPath` in a away that *does not* invoke `setFile` and the test will fail. THAT and ONLY THAT is what this test SHALL proof and nothing else.

Reply:
The use case he describes is NOT to be tested, as it is would not test the default method, but is about testing the actual implementing class. There is no need for such a test.